### PR TITLE
Read BOM via Path

### LINKED ISSUE #79: Read BOM from workspace submodule
Might be able to use

https://maven.apache.org/ref/3.5.4/maven-resolver-provider/apidocs/org/apache/maven/repository/internal/MavenWorkspaceReader.html

and Maven internal implementation of this. Maven doesn't expose a public API for this.

### LINKED ISSUE #521: Circle CI does not need to install Cloud OSS BOM
Once #485 is merged to master.

> If this works, we should be able to update .circleci/config.yml so it no longer manually installs the BOM before running tests.
https://github.com/GoogleCloudPlatform/cloud-opensource-java/pull/485

LinkageCheckerTest depends on the locally-installed BOM:

```
  @Test
  public void testGenerateInputClasspathFromLinkageCheckOption_mavenBom()
      throws RepositoryException, ParseException {
    String bomCoordinates = "com.google.cloud:cloud-oss-bom:pom:1.0.0-SNAPSHOT";

    LinkageCheckerArguments parsedArguments =
        LinkageCheckerArguments.readCommandLine("-b", bomCoordinates);
```

### PR DESCRIPTION
**Pull request by suztomo**
Fixes #79.

To discuss with Elliotte about the design when he comes back.

- RepositoryUtility.readBom now takes `Path` to a pom.xml
  This design does not require a local repository to have the BOM installed.
- DashboardMain.main to take a path to BOM as the argument.
  Kokoro's periodic build script now calls it with the argument

This `readBom` will help to run DashboardTest using a fixed pom.xml (#466).

Dependencies to locally-installed cloud-oss-bom has been removed. Fixes #521 .

### CODE REVIEW

File: dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java [hunk 1]
```diff
@@ -82,20 +86,31 @@
   public static final String TEST_NAME_GLOBAL_UPPER_BOUND = "Global Upper Bounds";
   public static final String TEST_NAME_DEPENDENCY_CONVERGENCE = "Dependency Convergence";
 
-  public static void main(String[] args)
-      throws IOException, TemplateException, RepositoryException, URISyntaxException {
-
-    Path output = generate();
+  /**
+   * Generates a code hygiene dashboard for a BOM. The file name (pom.xml) of the BOM is specified
```
Comment (elharo) [inline dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java]: I think this is a path, not just a file name.

File: dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java [hunk 2]
```diff
@@ -82,20 +86,31 @@
...
 
-  public static Path generate()
-      throws IOException, TemplateException, RepositoryException, URISyntaxException {
-
-    // TODO should pass in maven coordinates as argument
-    DefaultArtifact bom =
-        new DefaultArtifact("com.google.cloud:cloud-oss-bom:pom:1.0.0-SNAPSHOT");
-    List<Artifact> managedDependencies = RepositoryUtility.readBom(bom);
+  @VisibleForTesting
+  static Path generate(Path bomFile)
```
Comment (elharo) [inline dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java]: We might want a version that takes coordinates too.

File: dependencies/src/test/java/com/google/cloud/tools/opensource/dependencies/RepositoryUtilityTest.java [hunk 3]
```diff
@@ -41,12 +45,19 @@ public void testFindLocalRepository() {
...
     Assert.assertEquals(134, managedDependencies.size());
   }
-  
+
+  @Test
+  public void testReadBom_path() throws Exception {
+    Path pomFile = Paths.get("..", "boms", "cloud-oss-bom", "pom.xml");
+    ImmutableList<Artifact> artifacts = RepositoryUtility.readBom(pomFile);
+    Truth.assertThat(artifacts).isNotEmpty();
+    Truth.assertThat(artifacts.size()).isIn(Range.closed(150, 200));
```
Comment (elharo) [inline dependencies/src/test/java/com/google/cloud/tools/opensource/dependencies/RepositoryUtilityTest.java]: why not an exact number?

File: dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java (hunk 2 above)
Comment (suztomo) [reply dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java]: Added `generate(String)`

File: dependencies/pom.xml [hunk 4]
```diff
@@ -34,6 +34,16 @@
       <artifactId>bcel</artifactId>
       <version>6.2</version>
     </dependency>
+    <dependency>
+      <groupId>org.apache.maven</groupId>
+      <artifactId>maven-core</artifactId>
```
Comment (suztomo) [inline dependencies/pom.xml]: New dependency to implement `RepositoryUtility.createMavenProject`

File: dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java [hunk 5]
```diff
@@ -82,21 +86,41 @@
...
+   */
+  public static void main(String[] arguments)
+      throws IOException, TemplateException, RepositoryException, URISyntaxException,
+          PlexusContainerException, ComponentLookupException, ProjectBuildingException {
+    if (arguments.length != 1) {
+      System.err.println("Please specify path to pom.xml or Maven coordinates for a BOM.");
+      return;
+    }
+    String bomInput = arguments[0];
+    Path output = bomInput.endsWith("xml") ? generate(Paths.get(bomInput)) : generate(bomInput);
```
Comment (elharo) [inline dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java]: We should probably start using command line flags for this.

Comment (suztomo) [reply dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java]: Added command line parsing class `DashboardArguments`

File: .circleci/config.yml [hunk 6]
```diff
@@ -38,9 +38,6 @@ jobs:
 
       # run tests!
       - run: |
-          cd boms/cloud-oss-bom
```
Comment (suztomo) [inline .circleci/config.yml]: No locally-installed cloud-oss-bom

File: dashboard/pom.xml [hunk 7]
```diff
@@ -45,12 +45,6 @@
       <artifactId>dependencies</artifactId>
       <version>${project.version}</version>
     </dependency>
-    <dependency>
-      <groupId>com.google.cloud</groupId>
-      <artifactId>cloud-oss-bom</artifactId>
```
Comment (suztomo) [inline dashboard/pom.xml]: No locally-installed cloud-oss-bom.

File: dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassPathBuilderTest.java [hunk 8]
```diff
@@ -81,7 +81,7 @@ public void testArtifactsToPaths_removingDuplicates() throws RepositoryException
   @Test
   public void testBomToPaths_firstElementsAreBomMembers() throws RepositoryException {    
     DefaultArtifact bom =
-        new DefaultArtifact("com.google.cloud:cloud-oss-bom:pom:1.0.0-SNAPSHOT");
```
Comment (suztomo) [inline dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassPathBuilderTest.java]: This test cannot rely on locally-installed cloud-oss-bom.

File: dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerTest.java [hunk 9]
```diff
@@ -623,7 +623,7 @@ public void testFindClassReferences_privateClass() throws IOException, URISyntax
   @Test
   public void testGenerateInputClasspathFromLinkageCheckOption_mavenBom()
       throws RepositoryException, ParseException {
-    String bomCoordinates = "com.google.cloud:cloud-oss-bom:pom:1.0.0-SNAPSHOT";
```
Comment (suztomo) [inline dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerTest.java]: This test cannot rely on locally-installed cloud-oss-bom.

File: dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassPathBuilderTest.java [hunk 10]
```diff
@@ -81,7 +81,7 @@ public void testArtifactsToPaths_removingDuplicates() throws RepositoryException
...
+        new DefaultArtifact("com.google.cloud:google-cloud-bom:0.81.0-alpha");
```
Comment (suztomo) [inline dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassPathBuilderTest.java]: Now this test is fixed to this BOM. No need to update this test when we upgrade members of Cloud OSS BOM.

File: dashboard/src/test/java/com/google/cloud/tools/opensource/dashboard/DashboardTest.java [hunk 11]
```diff
@@ -121,10 +128,10 @@ private static Document parseOutputFile(String fileName)
   }
 
   @Test
-  public void testDashboard() throws IOException, ArtifactDescriptorException {
-    DefaultArtifact bom =
-        new DefaultArtifact("com.google.cloud:cloud-oss-bom:pom:1.0.0-SNAPSHOT");
```
Comment (suztomo) [inline dashboard/src/test/java/com/google/cloud/tools/opensource/dashboard/DashboardTest.java]: This test cannot rely on locally-installed cloud-oss-bom.

**Pull request merged**